docs/CMakeLists.txt

```diff
@@ -8,43 +8,51 @@
 #
 # SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
-include(FindPerl)
-if(PERL_FOUND)
+
+# These are all various expressions of the configuration schema
+get_filename_component(config_c_PATH "../src/core/ddsi/src/q_config.c" REALPATH BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
+
+# although the following files are generated, we check them into source control for convenience
+get_filename_component(config_rnc_PATH "../etc/cyclonedds.rnc" REALPATH BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
+get_filename_component(config_xsd_PATH "../etc/cyclonedds.xsd" REALPATH BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
+get_filename_component(config_md_PATH "./manual/options.md" REALPATH BASE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
+
+if(BUILD_SCHEMA OR BUILD_DOCS)
+  find_package(Perl REQUIRED)
   add_custom_command(
     OUTPUT
-      options.md cyclonedds.rnc
-    COMMAND
-      ${PERL_EXECUTABLE} -w "${CMAKE_CURRENT_SOURCE_DIR}/makernc.pl" "${CMAKE_CURRENT_SOURCE_DIR}/../src/core/ddsi/src/q_config.c" options.md cyclonedds.rnc
+    "${config_md_PATH}" "${config_rnc_PATH}"
     COMMAND
-      ${PERL_EXECUTABLE} -w "${CMAKE_CURRENT_SOURCE_DIR}/compare.pl" options.md "${CMAKE_CURRENT_SOURCE_DIR}/manual/options.md"
-    COMMAND
-      ${PERL_EXECUTABLE} -w "${CMAKE_CURRENT_SOURCE_DIR}/compare.pl" cyclonedds.rnc "${CMAKE_CURRENT_SOURCE_DIR}/../etc/cyclonedds.rnc"
+    ${PERL_EXECUTABLE} -w "${CMAKE_CURRENT_SOURCE_DIR}/makernc.pl" "${config_c_PATH}" "${config_md_PATH}" "${config_rnc_PATH}"
+    COMMENT "Generating Relax NG schema and Markdown documentation for config options"
     DEPENDS
-      "${CMAKE_CURRENT_SOURCE_DIR}/makernc.pl"
-      "${CMAKE_CURRENT_SOURCE_DIR}/../src/core/ddsi/src/q_config.c")
-  add_custom_target(options_doc ALL DEPENDS "options.md" "cyclonedds.rnc")
+    "${CMAKE_CURRENT_SOURCE_DIR}/makernc.pl"
+    "${config_c_PATH}")
+  add_custom_target(options_doc ALL DEPENDS "${config_md_PATH}" "${config_rnc_PATH}")
 
-  find_package(Java COMPONENTS Runtime)
-  if(JAVA_FOUND AND EXISTS "${TRANG_PATH}" OR EXISTS "$ENV{TRANG}")
-    if(NOT EXISTS "${TRANG_PATH}" AND EXISTS "$ENV{TRANG}")
-      message(STATUS "Setting TRANG_PATH to $ENV{TRANG}")
-      set(TRANG_PATH "$ENV{TRANG}" CACHE FILEPATH "Location of 'trang' for converting XML schemas" FORCE)
-    endif()
-    add_custom_command(
-      OUTPUT
-        cyclonedds.xsd
-      COMMAND
-        ${Java_JAVA_EXECUTABLE} -jar "${TRANG_PATH}" -I rnc -O xsd cyclonedds.rnc cyclonedds.xsd
-      COMMAND
-        ${PERL_EXECUTABLE} -w "${CMAKE_CURRENT_SOURCE_DIR}/compare.pl" cyclonedds.xsd "${CMAKE_CURRENT_SOURCE_DIR}/../etc/cyclonedds.xsd"
-      DEPENDS
-        "cyclonedds.rnc")
-    add_custom_target(options_xsd ALL DEPENDS "cyclonedds.xsd")
-  else()
-    message(STATUS "Java or not trang not found: not converting/checking RNC to XSD")
+  find_program(TRANG_EXECUTABLE trang)
+  if(NOT TRANG_EXECUTABLE)
+    message(FATAL_ERROR "Trang not found - needed for generating xsd files")
```

Comment:
@rotu you could also download trang when needed by doing

```
  set(TRANG_ZIP ${CMAKE_CURRENT_BINARY_DIR}/trang.zip)
  if(NOT EXISTS ${TRANG_ZIP})
    message(STATUS "Trang not detected: Downloading..")
    file(DOWNLOAD https://github.com/relaxng/jing-trang/releases/download/V20181222/trang-20181222.zip ${TRANG_ZIP})
    execute_process(COMMAND ${CMAKE_COMMAND} -E tar -xf ${TRANG_ZIP}
                    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
  endif()
  set(TRANG_EXECUTABLE ${CMAKE_CURRENT_BINARY_DIR}/trang-20181222/trang.jar)
```
However i think you want to make a cmake option if you want to download trang or not in the case that it is missing.

Reply:
1. A jar is not directly executable, at least on my Mac, so this would break:
```
$ cat `which trang`
#!/bin/bash
exec java  -jar /usr/local/Cellar/jing-trang/20181222/libexec/build/trang.jar "$@"

$ /usr/local/Cellar/jing-trang/20181222/libexec/build/trang.jar
zsh: permission denied: /usr/local/Cellar/jing-trang/20181222/libexec/build/trang.jar
```

2. Java is not necessarily installed either.

3. I think it's quite reasonable to expect build dependencies to be installed by the user or their build dependency management solution (e.g. `rosdep`).

Reply:
Would it not be an option to search for ``trang`` and if it doesn't exist, download the jar run it using ``java -jar``?

I rather like your changes (and dislike the thought of them getting abandoned because of a triviality like this). I guess I could also simply generate the XSD from my perl script: that has its advantages over being dependent on an external tool. I quickly abandoned that route in the past after seeing what an XSD looks like, but it is (self-evidently) just a matter of generating the right syntax.

Reply:
Downloading files and executing them from somewhere is a bad idea IMHO. If something like that is necessary, I would suggest doing that in the `.travis.yml`. It's really just like any other dependency(?)

Reply:
I don't love it, but I've implemented a cmake module to conditionally download Trang. Seems like a lot of code to save the user from a single `apt install trang`.

Reply:
Definitely like the CMake module. I think we should disregard @ThijsSassen's comment on downloading trang if it's missing, but stick with the module. Setting up a system should be done before building the software. At least, that's what I think. @ThijsSassen, is there a specific reason why we should download trang during the build and not when installing the system itself?

Reply:
Well the advantage of doing it in CMake is *all* builds of this project use CMake. On the other hand, some builds use `rosdep` (which doesn't even support `<doc_depend>` yet though it's in the package xml spec) some use `conan` (which does not have a listed version of `trang`), IDLC uses `Maven`.

It would be nice to have some saner dependency management, but ... https://xkcd.com/927/

Reply:
When I created the build system, I set it up with the basic rule that Conan is used for all C dependencies (libraries). Tools, like the compiler, Conan, CMake, Sphinx, etc must be available on the system beforehand. CMake is of course free to require a certain version, like you do in the module. It's just that when you mix building software and installing the system together, things will get messy. `rosdep` isn't used by Cyclone DDS (I believe) and Maven will eventually be removed when I get around to replace IDLC (Soonish, probably June/July)

Reply:
For the record, I'm sorry to only take part in this discussion now. It slipped past me. Other than the download bit the changes are a welcomed and very worthwhile addition. Thanks.

Reply:
Right now, Conan is *not* used for ROS2 builds. Only rosdep is, which uses the package.xml file. That may be the wrong call, but it's how things are expected for ROS2 where the workflow is generally:

`git clone`, `rosdep install`, `colcon build`

Rosdep is a frontend that dispatches to apt / pip / homebrew / chocolatey, etc. based on the platform and package. We could certainly use conan there, b

I agree that the download is non-ideal practice, but I just want to get these changes merged in :-(

Reply:
I'll let @eboasson, be the judge here as getting it in might be important enough to take this one and clean it up afterwards. If trang is required for the ROS2 build, I suggest specifying it in `package.xml` (can even be just a `depend`) and remove the download code from the CMake module. If it's not, list it as a `doc_depend` in `package.xml` and remove the download code from the CMake module. If it's required in the Eclipse Cyclone DDS builds on Travis, add it to `.travis.yml` too. That would be my preference.

Reply:
@k0ekk0ek i agree that when setting up the System would be the time to also install trang i.e. add it to the travis.yml as you suggested. As we currently have maven for java dependencies but as already mentioned maven does not provide trang i suggested to download it in cmake, but its also not my preferred method.

Reply:
@rotu, saw [this PR](https://github.com/ros/rosdistro/pull/24231). Much appreciated! Again, if this is required short-term, with a Chocolatey package not being available, I'm fine with it. We can worry about the proper solution later(?)

Reply:
Doesn't seem like rosdep is supported on Windows yet? Also, don't see a Chocolatey package installer implementation for rosdep, so Windows may not be a problem(?) That being said, I do see a SourceInstaller Python module there. @rotu, can that installer module be used to download the archive and install it for ROS2 or would we need an actual Chocolatey package? Anyway, just curious. If you have some information, let me know, I might be able to help out.

Reply:
Took the liberty to create a [patch for `.travis.yml`](https://gist.github.com/k0ekk0ek/8a53f708eb098cd9c4006bbb674e6c24).

Reply:
`rosdep` [does not yet support `doc_depend`](https://www.ros.org/reps/rep-0149.html#doc-depend-multiple), so even if Trang is available for all target platforms and rosdep works perfectly on Windows, it's not going to work yet. I added `trang` to the package.xml more for documentation than for functionality at this point in time.

I don't want to throw any more time down this hole right now. Please come to an agreement with @ThijsSassen and @eboasson and I'll gladly code it up later whichever way you guys like.

Reply:
We (@eboasson, @ThijsSassen and I) are in agreement that downloading is definitely not the way to go. Once that's removed, I'm happy to merge this in. Thanks.

Reply:
@k0ekk0ek Done.

Reply:
Didn't go unnoticed @rotu. Actually had a quick look, but lost myself in trying to find a way to ditch Perl and Java. Not important for this PR, I'll have another look later today.

Reply:
I support that!